worker: initial implementation

addaleax opened the pull request:
edit: moved to https://github.com/ayojs/ayo/pull/58https://github.com/ayojs/ayo/pull/58

(status: currently ready for review, by anyone, including you: https://github.com/ayojs/ayo/pull/40#issuecomment-328358638)

```js
'use strict';
const { Worker } = require('worker');

if (process.isMainThread) {
  module.exports = async function parseJSAsync(script) {
    return new Promise((resolve, reject) => {
      const worker = new Worker(__filename, {
        workerData: script
      });
      worker.on('message', resolve);
      worker.on('error', reject);
      worker.on('exit', (code) => {
        if (code !== 0)
          reject(new Error(`Worker stopped with exit code ${code}`));
      });
    });
  };
} else {
  const { parse } = require('some-js-parsing-library');
  const script = process.workerData;
  process.postMessage(parse(script));
}
```

(If that gives the wrong impression: No, this does not conform the browser WebWorker API, but that should be rather easily implementable on top of this, and I’m okay with that.)

Fixes: https://github.com/ayojs/ayo/issues/31

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### TODO

- [x] maybe turn the message passing mechanisms into standard [MessageChannel](https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel)s
- [x] implement ArrayBuffer transferring
- [x] implement SharedArrayBuffer, uh, Sharing
- [x] implement MessagePort transferring
- [x] remove the need to call `.start()` manually for `MessagePort`s
- [x] add tests for MessageChannels
- [ ] improve uncaught exception serialization/deserialization
- [ ] fix garbage collection tracking through `performance` (currently pending the V8 6.1 update in upstream Node.js)
- [ ] add async hooks tests for Workers
- [x] figure out how to best run as many parallel/ tests as possible from Workers (did that by basically taking @petkaantonov’s approach directly from https://github.com/petkaantonov/io.js/commit/ea143f72fc6845668716bd7a5da22771011defe4)
- [ ] figure out whether and how this can be integrated with the inspector
- [x] figure out a native addon story
- [x] make deserialization context for `MessagePort`s configurable
- [x] look out for memory leaks created by node internals not cleaning up on environment destruction
- [ ] integrate `v8::Platform` implementation with multi-isolate support

(@petkaantonov please feel free to indicate whether attributing you for code that comes from your original PR is not enough/too much/just right :) )

petkaantonov committed:
test: add more tests for workers

Taken from https://github.com/petkaantonov/io.js/commit/ea143f72fc6845668716bd7a5da22771011defe4
and modified to fit current linter rules and coding style.

addaleax committed:
worker: initial implementation

addaleax committed:
console: make global console work in workers

benjamingr commented on lib/domain.js:
```diff
@@ -21,6 +21,10 @@
 'use strict';
 
+if (process.isWorkerThread) {
+  throw new Error('domains are not available inside of worker threads');
```
That's a pretty big decision to make. Personally I'm +1 on it but it means that any code that currently runs in child processes won't be able to run on workers.

Petka's original idea was to convert `cluster` to use workers.

addaleax replied:
> it means that any code that currently runs in child processes won't be able to run on workers.

You mean “some code”, right? As in, child processes and workers are incompatible? I’m okay with that as well.

> Petka's original idea was to convert `cluster` to use workers.

I would assume by now too much code relies on that not being the case anyway.

benjamingr replied:
> You mean “some code”, right? As in, child processes and workers are incompatible? I’m okay with that as well.

That sentence was missing "and uses domains" :)

I think that people currently using child processes to offload work are the most obvious potential consumers of workers and this _might_ make their life harder.

I'm definitely OK with starting this way in any case.

> I would assume by now too much code relies on that not being the case anyway.

I'm not sure why?

addaleax replied:
> > I would assume by now too much code relies on that not being the case anyway.

> I'm not sure why?

For one, the current implementation in this PR doesn’t provide stdio to Workers, but I would be pretty sure a lot of cluster-using code wants that to be there ;)

benjamingr replied:
Come to think of it - can't stdio just be done through a MessagePort?

addaleax replied:
@benjamingr Yes … but that could also be done by the user, right? `stdio` in Node/Ayo is complicated, and I’m a bit scared of what happens when multiple handles to the same resource exist (like, some could be set to blocking, data would end up being intertwined, there would be different libuv buffers, and for `stdin` there can’t really be any consistent behaviour anyway.

I’ve made `console.*` work by just using a fully synchronous stream that operates directly on the file descriptors, which might not be super-efficient but I think it’s reasonable here.

benjamingr replied:
I think that's reasonable and we can always add it later.

benjamingr commented on lib/internal/worker.js:
```diff
@@ -0,0 +1,275 @@
+    debug(`[${process.threadId}] created Worker with ID ${this.threadId}`);
+
+    // Actually start the new thread now that everything is in place.
+    this[kHandle].startThread();
```
I'm not sure if we shouldn't provide `.start()` as an external method so creating and starting the work are different, but I don't feel strongly about it since the user can always just keep a function returning a thread.

addaleax replied:
Yeah, it’s an arbitrary decision in my eyes, too. Right now, all communication with the Worker is asynchronous anyway, so I couldn’t find any advantage in adding a separate method, but if there is some, fine by me. :)

benjamingr commented on lib/internal/worker.js:
```diff
@@ -0,0 +1,275 @@
+
+const assert = require('assert');
+const errors = require('internal/errors');
+const { Serializer, Deserializer } = require('v8');
```
We need a way to pass binary data fast between isolates - given we can't pass structured data (right?) I think that we need a way to pass buffers and other node builtins fast if we can.

TimothyGu replied:
The custom Serializer does support Node.js Buffers. Are you saying the performance of going through the serializer might be suboptimal?

benjamingr replied:
I'm saying that passing binary data efficiently between workers needs to be fast, it would be amazing to be able to pass objects faster than structured-cloning them but I don't see that happening any time soon.

benjamingr replied:
One of the most commonly misunderstood problems people want to solve with workers is "offload parsing a lot of JSON" where the overhead is creating all the objects and not reading the actual JSON and passing it to a worker won't help (since now 2 copies are made per worker because it's deserializaed from JSON to object, then between workers, then to the user).

I don't think we can fix that though.

addaleax replied:
@benjamingr Maybe to clarify: Are you saying the serialization (that corresponds to structured cloning, and afaik is what Chrome implements structured cloning on top of) needs to be fast, or that passing in a Buffer on the one side and getting out another Buffer on the other side needs to be fast?

Anyway:
- I think we might be able to get rid of one (?) of the copy operations along the way when a Buffer is passed in by adding a fast case for “this is just an ArrayBufferView”
- I would also like to add `SharedArrayBuffer`/transferable support, now or in the future

> I don't think we can fix that though.

Yeah.

benjamingr replied:
> that passing in a Buffer on the one side and getting out another Buffer on the other side needs to be fast?

That, I think that passing a buffer should be fast (and optimally a binary stream). I have no hopes for anything that requires actually copying JS objects from one heap to another.

addaleax replied:
We won’t be able to get around one of the copies, because any `Buffer` that’s being passed in could be modified in the sending thread afterwards, but I think keeping it at a single copy should work.

TimothyGu replied:
@benjamingr Yeah. A FAQ/explainer for questions like these should be good, though that might sound too much like a Node.js EP... *sigh*

benjamingr replied:
@addaleax I'm fine with modifying a sent buffer `throw`ing or providing a different type.

I'm not too concerned with the API but I definitely think workers should have the capability to pass memory without copies _somehow_.

For example, if I'm using workers to transcode video and I'm doing it at 20Mbps in 4 workers - that means my process is copying 4.8 gigabits of memory every minute which is a lot of overhead.

addaleax replied:
> I'm fine with modifying a sent buffer `throw`ing or providing a different type.

Oh, right. I think we can `Neuter()` array buffers like that (or even have to do that when using them as transferrables?).

> I'm not too concerned with the API but I definitely think workers should have the capability to pass memory without copies _somehow_.

I think that would be doable with SharedArrayBuffers in any case

(to be clear, leaving out support for SABs/transferrables is not a design decision of this PR, it’s a `TODO` of it :D)

TimothyGu commented:
Awesome work!

I do have some questions about this implementation. It doesn't have to be as detailed (or contentious) as a Node.js EP, but I think these questions are likely to come up in the future. Please understand that, by asking these questions, I'm also genuinely interested in the answers and not trying to pass any judgements on the decisions made by this implementation.

1. ESM support. With ESM being an active WIP, and something that is spec'd for Web Worker, does this design for Worker allow future extension to ESM?
1. How much more work does implementing the Web Worker API on top of this API necessitate?
1. Blackboxing globals. A use case for Workers identified earlier is for execution of untrusted code. This impl already provided memory constraints, but maybe an option should be provided to not grant the Worker any special access to the system except to, well, work? (i.e. no require, no module, no process, only message posting, which i guess does sound like Web Workers)
1. Resource limits. As mentioned earlier, memory constraints are a thing in this impl. Is it possible to expose more system limits (e.g. ulimit) in this API in the future?
1. Synchronous IO. I remember @domenic mentioned synchronous IO as a potential use case for Workers, and sync XHR is indeed a thing in Web Workers. Is it possible to extend synchronocity to more core modules with something like this API in place?
1. SharedArrayBuffer. I'd like to confirm that this can be implemented, as it's one of the more exciting things in ES recently. Right?
1. Collaboration with Node.js. I do understand the rationale for posting the PR here, and I fully agree with them. But the current fact is that a lot of people capable of reviewing this PR properly have not yet hopped on Ayo.js. I'd like this PR to draw more people into Ayo.js, but IMO we should be thinking about meeting them in the middle.

EDIT: Added two more questions.

benjamingr commented:
Very nice questions, I looked at the code and Anna will probably have more to add on top of it - but doing the best of my ability to answer:

> ESM support. With ESM being an active WIP, and something that is spec'd for Web Worker, does this design for Worker allow future extension to ESM?

I don't see why not. This design (if I understand correctly) isn't interested with module loading semantics.

> How much more work does implementing the Web Worker API on top of this API necessitate?

Not a lot at all - I'm not sure that's a goal though.

> Blackboxing globals. A use case for Workers identified earlier is for execution of untrusted code.

To be honest that's something that I think should be solved by using the `vm` module inside a worker rather than making a worker do two things. I think if web had frozen realms (https://github.com/tc39/proposal-realms) when WebWorkers landed that's what they would have done too.

> Synchronous IO. I remember @domenic mentioned synchronous IO as a potential use case for Workers, and sync XHR is indeed a thing in Web Workers. 

I think CPU intensive work on in-process memory is a much more important goal for web workers. I think as a project Ayo (or Node.js) shouldn't provide any more tools for synchronous I/O since libuv already thread-pools blocking I/O operations like DNS lookups and they look "as asynchronous" to the user.

I'll let Domenic answer what he meant though :) (note he has asked to not be pinged in Node.js repos, I'm not sure how/if that holds here).

addaleax commented:
> I'm also genuinely interested in the answers and not trying to pass any judgements on the decisions made by this implementation.

You’re not coming across any other way :) Also, it’s fair to question my design decisions, they *are* somewhat ad-hoc and I wouldn’t be surprised if this PR did a shift in some other direction before it lands.

And @benjamingr’s answers are good answers :)

> Resource limits. As mentioned earlier, memory constraints are a thing in this impl. Is it possible to expose more system limits (e.g. ulimit) in this API in the future?

I think `ulimit` does always work on a per-process level, so I would go with “probably not”?

TimothyGu commented:
> I think `ulimit` does always work on a per-process level, so I would go with “probably not”?

Hmm, right.

I was also more wondering more generally about per-thread resource limits, so not necessarily those exposed through ulimit. I.e. more in the sense of "what else can we do with Workers?"

addaleax commented:
> I was also more wondering more generally about per-thread resource limits, so not necessarily those exposed through ulimit. I.e. more in the sense of "what else can we do with Workers?"

If you’re thinking about limiting a Worker’s access to APIs, I’d go with the combination of a Worker + a new VM context that Benjamin mentioned, and let userland handle that case from there.

Otherwise, I think this really depends on what kind of limits you’re having in mind?

addaleax committed:
test: add basic tests for worker

addaleax commented:
I’ve added a TODO list in the PR description, if anybody has interest in jumping in on something specific, please let me know (here, or if that doesn’t work, ping me in the discord channel/twitter/wherever)

addaleax commented:
@timothygu for your more recent questions:

> SharedArrayBuffer. I'd like to confirm that this can be implemented, as it's one of the more exciting things in ES recently. Right?

Yes. V8 already has an implementation of proper message passing code that handles transferring ABs and sharing SABs through workers in `d8`, you can take a look at that if you like

> Collaboration with Node.js. I do understand the rationale for posting the PR here, and I fully agree with them. But the current fact is that a lot of people capable of reviewing this PR properly have not yet hopped on Ayo.js. I'd like this PR to draw more people into Ayo.js, but IMO we should be thinking about meeting them in the middle.

Just to be clear, my personal rationale is more that I’m tired and frustrated with Node’s review process. That this would give Ayo a technical edge over Node is really really nice, but not the main issue for me.

I would hope that, if done right, this could also be a good chance to give people who have *not* felt welcome around Node.js to enter the scene. I would be more than happy to answer questions newcomers have about this PR, whether very abstract or very concretely referring to pieces of code.

If that fails, yes, we can go back to looking for Node people.

(Also, to clarify: Do you consider yourself a person “capable of reviewing this PR”? I personally would trust your review on this).

addaleax committed:
src: keep track of env properly in node_perf.cc

Currently, measuring GC timing using `node_perf` is somewhat broken,
because Isolates and Node Environments do not necessarily match 1:1;
each environment adds its own hook, so possibly the hook code runs
multiple times, but since it can’t reliably compute its corresponding
event loop based on the Isolate, each run targets the same Environment
right now.

This fixes that problem by using new overloads of the GC tracking
APIs that can pass data to the callback through opaque pointers.

TimothyGu commented:
> Just to be clear, my personal rationale is more that I’m tired and frustrated with Node’s review process. That this would give Ayo a technical edge over Node is really really nice, but not the main issue for me.
>
> I would hope that, if done right, this could also be a good chance to give people who have not felt welcome around Node.js to enter the scene. I would be more than happy to answer questions newcomers have about this PR, whether very abstract or very concretely referring to pieces of code.
>
> If that fails, yes, we can go back to looking for Node people.

Thank you for this rationale for posting the PR here. It makes total sense, though I agree we'll have to wait and see regarding the reviews we will get.

> (Also, to clarify: Do you consider yourself a person “capable of reviewing this PR”? I personally would trust your review on this).

To a certain extent, yes, I do. On the other hand, I would not feel comfortable signing off on this if this was to go in w/o other more capable pairs of eyes over it :)

addaleax committed:
src: make CleanupHandles() tear down handles/reqs

addaleax committed:
src: use lock for c-ares library init/cleanup

addaleax committed:
src: protect global state with mutexes

addaleax committed:
worker: implement `MessagePort` and `MessageChannel`

addaleax commented:
Fwiw, I just updated the `worker` module to include mostly-standard [`MessageChannel`](https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel) support, so if anybody wants to write tests for that, please write your hand :)

(There are no docs for that yet, but the API matches the MDN descriptions except that events are emitted via a Node.js-style `message` event instead of `.onmessage`)

aikar commented:
Why is this its own standalone API? I understand the goal is to use threads for improved data passing, but IMO this absolutely should be an extension of the cluster API, in a sense that if cluster is simply reforking itself or another node script, that it could be an option to use a thread instead of child process.

Then your not designing any new API. Code works just the same, but with a potential performance improvement.

If you design it as you are, good luck with adoption as regardless of peoples thoughts on why this ayo project was started, people aren't going to drop support for node.js, so they aren't going to use an API that won't work in node.

But if cluster.fork() had an option threads: true, that enabled this alternate behavior, that property would be ignored on node.js and then function the same on both platforms.

addaleax committed:
test: add basic tests for MessageChannel

Qard commented:
@aikar 

1. Cluster and workers have fundamentally different purposes and scope. It doesn't make sense to limit the capabilities to the API currently used for clustering.
2. This PR is an attempt to bring the WebWorker spec from the browser to Node.js, so the fact it doesn't exist in Node.js is irrelevant. This is about server/browser parity.
3. The threads option for cluster could probably be implemented on top of this.
4. This is a WIP, meaning this work is incomplete. There's still tons of stuff that will likely change. What you see now will probably become the core bits of whatever the final result is.

addaleax commented:
> This PR is an attempt to bring the WebWorker spec from the browser to Node.js, so the fact it doesn't exist in Node.js is irrelevant. This is about server/browser parity.

If I may add to that: This PR doesn’t seek to implement the `WebWorker` spec. It does seek to make that spec implementable as a userland module, though.

> The threads option for cluster could probably be implemented on top of this.

Probably – and probably also doable in userland.

addaleax commented:
> If you design it as you are, good luck with adoption as regardless of peoples thoughts on why this ayo project was started, people aren't going to drop support for node.js, so they aren't going to use an API that won't work in node.

That is also kind of true for new APIs that are being introduced in Node – adoption will be gradual either way, and I would expect to see modules have optional support for it before they fully depend on it.

Qard commented:
Ah, yeah. Good point. Thanks for clarifying. Lower-level capabilities are better, as it's much more flexible that way. 👍

addaleax commented:
Update: It appears that, among other things, `SharedArrayBuffer`s are working now :) The feature is behind V8’s `--harmony-sharedarraybuffer` flag¹, and I realize this PR still lacks documentation, but I invite everyone who wants to give it a try to do so.

This is the simple test case that checks basic functionality and demonstrates how to use the API: https://github.com/addaleax/ayo/blob/5724e3ba444451d73a6c72ad8da1dda34ea52f4a/test/parallel/test-message-channel-sharedarraybuffer.js

¹ edit: if you’re curious, it’s enabled by default in V8 6.2, which should become stable during mid-October

addaleax committed:
deps: cherry-pick b6158eb6befae from V8 upstream

Original commit message:

    [heap] Move gc callbacks from List to std::vector

    Bug: v8:6333
    Change-Id: I4434c6cc59f886f1e37dfd315a3ad5fee28d3f63
    Reviewed-on: https://chromium-review.googlesource.com/634907
    Reviewed-by: Ulan Degenbaev <ulan@chromium.org>
    Commit-Queue: Michael Lippautz <mlippautz@chromium.org>
    Cr-Commit-Position: refs/heads/master@{#47601}

addaleax committed:
deps: cherry-pick 9b21865822243 from V8 upstream

Original commit message:

    [api] Add optional data pointer to GC callbacks

    This can be useful when there may be multiple callbacks attached by
    code that's not directly tied to a single isolate, e.g. working
    on a per-context basis.

    This also allows rephrasing the global non-isolate APIs in terms
    of this new API, rather than working around it inside `src/heap`.

    TBR=hpayer@chromium.org

    Bug:
    Cq-Include-Trybots: master.tryserver.chromium.linux:linux_chromium_rel_ng
    Change-Id: I2e490ec40d1a34ea812f25f41ef9741d2116d965
    Reviewed-on: https://chromium-review.googlesource.com/647548
    Reviewed-by: Yang Guo <yangguo@chromium.org>
    Reviewed-by: Adam Klein <adamk@chromium.org>
    Commit-Queue: Yang Guo <yangguo@chromium.org>
    Cr-Commit-Position: refs/heads/master@{#47923}

addaleax committed:
src: add environment cleanup hooks

addaleax committed:
src: use cleanup hooks to tear down BaseObjects

addaleax committed:
addons: integrate workers with native addons

Native addons need to use flags to indicate that they are capable
of being loaded by worker threads.

Native addons are unloaded if all Environments referring to it
have been cleaned up, except if it also loaded by the main Environment.

addaleax committed:
worker: implement vm.moveMessagePortToContext()

This should help a lot with actual sandboxing of JS code.

addaleax committed:
src: rename string id for `_onclose` to match

addaleax committed:
async_wrap: update utility methods for flexibility

addaleax committed:
doc: add documentation for workers module

addaleax commented:
There are a few minor outstanding items on the list, but generally, what is here is ready to be reviewed. :tada: 

Some notes and suggestions:
- You can take looks at individual commits, and approving or commenting individual commits (e.g. through comments) *is* helpful – whether you are a collaborator or not.
- The docs commit (currently the last one) is probably a good starting point.
- I appreciate *anybody’s* review or questions, *especially* including anything that feels like a “dumb” question.
- There is quite a bit of C++ and shifting around internals of Node involved. Literally nobody can be expected to be familiar with all of it, so again, questions are very very cool, including questions that are in the general direction “how does this particular Node thing/V8 thing/… work”, even if it’s not something that’s changed in this PR.
- You can contact me in the discord (channel or dm) about anything related to this PR.

addaleax commented:
I’m mostly busy for the remainder of the week, but we might need to figure out sth for reviews here.

I realize this is a pretty huge PR for a first technical feature of our own, but realistically, for this to happen we’re going to need volunteers to review parts of it (and again, just looking at the doc or test commits is helpful in reducing that work).

I think for the non-C++ parts that *should* be doable given how many people indicate that they want to help with this project. For the C++ parts that might get a bit harder, but I’ll be glad to put a bit of work into making that happen as well (e.g. do bits of code review in a hangout session with interested people so it’s easier to ask questions? wdyt?).

benjamingr commented on doc/api/domain.md:
```diff
@@ -25,6 +25,8 @@ will be notified, rather than losing the context of the error in the
 `process.on('uncaughtException')` handler, or causing the program to
 exit immediately with an error code.
 
+*Note*: This module is not available in [`Worker`][]s.
```
So to clarify, if a worker throws what should happen?

addaleax replied:
You get an `error` event in the main thread, which I now realize is missing from the documentation.

benjamingr commented on doc/api/process.md:
```diff
@@ -1365,6 +1394,18 @@ system platform on which the Node.js process is running. For instance
+added: REPLACEME
+-->
+
+This method is available if this is a `Worker` thread, which can be tested
```
I'm not sure if all the worker stuff should go on `process` and not on the `Worker` object. `process` is already really big and `process.postMessage` sounds a little confusing.

addaleax replied:
You mean, make all of these statics on the `Worker` constructor? Or the `worker` module? I’m okay with that, it doesn’t really make a difference and I agree that `process` is already pretty big.

benjamingr replied:
The worker module and the constructor both sound fine

benjamingr commented on doc/api/process.md:
```diff
@@ -1824,6 +1890,14 @@ cases:
+added: REPLACEME
+-->
+
+An arbitrary JavaScript object that contains a clone of the data passed
```
"contains a clone" could use clarifying - for example what happens to functions on the object.

addaleax replied:
Yeah, I’m linking the HTML structured clone algorithm here as well

benjamingr commented on doc/api/worker.md:
```diff
@@ -0,0 +1,255 @@
+> Stability: 1 - Experimental
+
+The `worker` module provides a way to create multiple environments running
+on independent threads, and to create message channels between those. It
```
Might want to mention sharing memory as something this solves.

benjamingr commented on doc/api/worker.md:
```diff
@@ -0,0 +1,255 @@
+added: REPLACEME
+-->
+
+The `MessageChannel` has no methods of its own. `new MessageChannel()`
```
Are these compatible with DOM message channels? If so we should link to the DOM spec on those here somewhere.

addaleax replied:
Almost – this is doing `MessageChannel`s as Node `EventEmitter`s rather than `EventTarget`s, but otherwise it should work the same

benjamingr reviewed (commented):
Some more thoughts

addaleax commented:
Okay, resolved conflicts + got CI back to passing again after the upstream update.

Qard commented:
Just playing with it a bit right now. One thing that might trip up new users a bit is that a `Buffer` becomes a `Uint8Array` in the transfer. One might naively try to pass a buffer representing text, like file contents, into a worker and expect to call `toString()` on it at the other end, but that'll produce the array-style comma-separated list of numbers rather than the string they might be expecting.

Not sure if there's a way to automatically restore the buffer type on the other end or if it should just be mentioned in the docs that this happens.

Qard commented:
Just got `AssertionError [ERR_ASSERTION]: An invalid error message key was used: ERR_WORKER_OFFLINE.` while playing with an http server in a worker.

https://gist.github.com/Qard/b0928fb4e92f53d4789702cd8ed2641a

addaleax committed:
src: prepare v8 platform for multi-isolate support

addaleax commented:
> One thing that might trip up new users a bit is that a `Buffer` becomes a `Uint8Array` in the transfer.

Yes – that is tricky. On the one hand, if we want to follow the Web’s structured cloning algorithm, we should stick to it, and we’re supporting serialization between `vm` contexts here which just don’t always have `Buffer`. Plus, most of Node’s API doesn’t make a difference between `Buffer`s and `Uint8Array`s anymore anyway.

On the other hand, I get that people might expect `Buffer` → `Buffer` serialization, and we should probably warn against sending buffers created with `Buffer.allocUnsafe()` or `Buffer.allocUnsafeSlow()` that might contain privileged information …

> Just got `AssertionError [ERR_ASSERTION]: An invalid error message key was used: ERR_WORKER_OFFLINE.` while playing with an http server in a worker.

Yeah… my bad. :) I’ll align the behaviour with what WebWorkers do (i.e. just ignore posted messages/terminate requests for already-terminated workers)

addaleax commented:
Moving to https://github.com/ayojs/ayo/pull/58 after we renamed branches
